kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py

```diff
@@ -0,0 +1,220 @@
+        self.errorCovPost = self.get_parameter("errorCovPost")._value
+        self.a_noise = self.get_parameter("a_noise")._value
+        self.vel_filter = self.get_parameter("vel_filter")._value
+        self.top_down = self.get_parameter("top_down")._value
+        self.cost_filter = self.get_parameter("cost_filter")._value
+
+        self.obstacle_list = []
+        self.max_id = 0
+        self.sec = 0
+        self.nanosec = 0
+
+        # subscribe to detector 
+        self.detection_sub = self.create_subscription(
+            ObstacleArray,
+            "detection",
+            self.callback,
+            10)
+
+        # publisher for tracking result
+        self.tracker_obstacle_pub = self.create_publisher(ObstacleArray, 'tracking', 10)
+        self.tracker_marker_pub = self.create_publisher(MarkerArray, 'marker', 10)
+
+    def callback(self, msg):
+        '''callback function for detection result'''
+
+        # update delta time
+        dt = (msg.header.stamp.sec - self.sec) + (msg.header.stamp.nanosec - self.nanosec) / 1e9
+        self.sec = msg.header.stamp.sec
+        self.nanosec = msg.header.stamp.nanosec
+
+        # get detection
+        detections = msg.obstacles
+        num_of_detect = len(detections)
+        num_of_obstacle = len(self.obstacle_list)
+
+        # kalman predict
+        for obj in self.obstacle_list:
+            obj.predict(dt)
+
+        # hungarian matching
+        cost = np.zeros((num_of_obstacle, num_of_detect))
+        for i in range(num_of_obstacle):
+            for j in range(num_of_detect):
+                cost[i, j] = self.obstacle_list[i].distance(detections[j])
+        obs_ind, det_ind = linear_sum_assignment(cost)
+
+        # filter assignment according to cost
+        new_obs_ind = []
+        new_det_ind = []
+        for o, d in zip(obs_ind, det_ind):
+            if cost[o, d] < self.cost_filter:
+                new_obs_ind.append(o)
+                new_det_ind.append(d)
+        obs_ind = new_obs_ind
+        det_ind = new_det_ind
+
+        # kalman update
+        for o, d in zip(obs_ind, det_ind):
+            self.obstacle_list[o].correct(detections[d])
+
+        # birth of new detection obstacles and death of disappear obstacle
+        self.birth(det_ind, num_of_detect, detections)
+        dead_object_list = self.death(obs_ind, num_of_obstacle)
+
+        # check if there is subscirbers
+        if self.tracker_obstacle_pub.get_subscription_count() == 0 and self.tracker_marker_pub.get_subscription_count() == 0:
+            return
+
+        # construct ObstacleArray
+        obstacle_array = ObstacleArray()
+        obstacle_array.header = msg.header
+        track_list = []
+        for obs in self.obstacle_list:
+            # do not publish obstacles with low speed
+            obs_vel = np.linalg.norm(np.array([obs.msg.velocity.x, obs.msg.velocity.y, obs.msg.velocity.z]))
+            if obs_vel > self.vel_filter:
```

Comment:
We should have min and max speeds, that way if there's a bogus correlation so the velocity appears to be physically impossible we can catch that.